### ISSUE TITLE
Is request body no longer allowed in DELETE request?

### ISSUE THREAD

**Issue by nexdrew** [sindresorhus/got#1028]
#### What would you like to discuss?

Is there a way to send a request body in a DELETE request?

Similar to #966 but specific to a DELETE request.

This was possible in got 9 like so:

```js
got(url, {
  method: 'DELETE',
  json: true,
  body: { version: 0 }
})
```

But attempting this in got 10 does not transmit a request body:

```js
got(url, {
  method: 'DELETE',
  responseType: 'json',
  json: { version: 0 }
})
```

I attempted to use `json` and `body` (via `JSON.stringify()`), with and without specifying a `Content-Type: application/json` header.

Am I doing it wrong?

I don't see any mention of this change in the [v10 release notes](https://github.com/sindresorhus/got/releases/tag/v10.0.0).

The best I can find about this in any RFC is [here](https://tools.ietf.org/html/rfc7231#section-4.3.5):

> A payload within a DELETE request message has no defined semantics; sending a payload body on a DELETE request might cause some existing implementations to reject the request.

which does not seem to expressly forbid it.

#### Checklist

- [x] I have read the documentation.

**Comment by KSDaemon** [comment]
May i put my 5 coins: why not just give it up to developer? If someone (like me or @nexdrew) wants to send some payload — do it. If it's a problem — just send nothing.

**Comment by sindresorhus** [comment]
I don't think we did anything to explicitly prevent this. It's probably just a case of us not anticipating this use-case, so we didn't have tests for it, and it accidentally regressed.

It would be helpful if someone could submit a pull request with a [failing tests](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#failing-tests). That would help getting this fixed faster.

**Comment by nexdrew** [comment]
Please see #1031.

**Comment by jrnelson90** [comment]
After poking around, it seems to be revolve around logic moved into `normalize-arguments.ts` from `request-as-event-emitter.js` in got 9.6. This logic is related to specifying a content-length header value when a payload body is present on a request.


https://github.com/sindresorhus/got/blob/d222164dffeb3ebf7d3ecb710fd56e2af5324c81/source/normalize-arguments.ts#L396-L413

You can see that this being handled for the cases POST, PUT, and PATCH, but not DELETE. Adding the case for DELETE seems to allow my requests to get 200 response instead of a 400 for having an empty body. That said, it did not seem to make the new test from @nexdrew pass.

**Comment by jrnelson90** [comment]
Here's a temporary workaround solution (based on @nexdrew's original post) that seems to work with the current npm version, **got 10.2.2**:
```
const jsonBody = { version: 0 };
got(url, {
  method: 'DELETE',
  responseType: 'json',
  json: jsonBody,
  headers: {
    "content-length": String(Buffer.byteLength(JSON.stringify(jsonBody))),
  },
});
```

**Comment by nexdrew** [comment]
Concerning this:

> You can see that this being handled for the cases POST, PUT, and PATCH, but not DELETE. Adding the case for DELETE seems to allow my requests to get 200 response instead of a 400 for having an empty body. That said, it did not seem to make the new test from @nexdrew pass.

I concur. My test is bad. I failed to notice that the Express server created by `create-test-server` is configured to disable the default body parsing options here: https://github.com/sindresorhus/got/blob/d222164dffeb3ebf7d3ecb710fd56e2af5324c81/test/helpers/with-server.ts#L17-L19

The test I wrote assumes that `content-type: application/json` would be parsed by the test server, but it is not, so the response body will always be empty (an empty object).

If you replace my dumb endpoint handler here: https://github.com/sindresorhus/got/blob/d222164dffeb3ebf7d3ecb710fd56e2af5324c81/test/post.ts#L266-L270

with this:

```js
server.delete('/', defaultEndpoint);
```

Then the test passes.

@jrnelson90 Good sleuthing, thanks for the feedback! You found the appropriate fix in `normalize-arguments.ts`.

Unless someone beats me to it, I will open a PR to (a) fix my test and (b) fix sending a request body for a DELETE, per Justin's findings.
